content/posts/intuition.md

```diff
@@ -0,0 +1,37 @@
+
+To overcome these difficulties we use two different and complementary methods. One of them is TLA+, a framework for proving correctness of distributed algorithms. We will write more about TLA+ and our usage of it in one of the future posts. This time we want to focus on the other method -- our intuition. Over past months we developed a set of heuristics that are very helpful in sketching the structure of an algorithm. They are useful in initial phases of research before a full TLA+ model is developed. We'd like to introduce you some of them.
+
+### The transaction and the side effects
+
+The outcome of processing a message consists of two parts. There is a transactional part and a side effects part. The transaction consists of application state change and of marking the incoming message as processed. The side effects include things like creating objects in non-transactional data stores (e.g. uploading a blob) and sending messages.
+
+### Until the transaction is committed, nothing happened
+
+In order for an algorithm to behave correctly, it has to guarantee that until a transaction is committed, no effects of the processing are visible to the external observers. Violation of that rule results in the ghost message or ghost data phenomenons. Ghost messages are messages that carry information about a state change that has never occurred.
+
+### Prepare - Commit - Publish
+
+So we don't want to create any public information before the state change is committed. Bus we also don't want to end up in a situation where we did commit the transaction but failed right after that and lost the information we were about to announce to the world. For this reason any correct algorithm has to make sure the side effects are made durable, but not visible, before the transaction is committed. Then, after the commit, the side affects are published.
+
+### Side effects stored in each processing attempt are isolated
+
+This one is better explained using an example. Let's say our processing consists of generating a PDF, storing it in a blob store and sending out a message that contains a SHA hash of that document. The generation of the PDF document is not 100% deterministic due to a timestamp that is rendered in the top right corner. If we allow the PDF to be stored in a deterministic location, same for each processing attempts, we may end up in a situation where the message that is sent out contains a hash of a document that has been overridden by another processing attempt. This may happen is the incoming message has been duplicated. One copy has been processed correctly while the other copy succeeded to complete the Prepare phase but failed to commit and was subsequently discarded as a duplicate.
```

Comment:
```suggestion
This one is better explained using an example. Let's say our processing consists of generating a PDF, storing it in a blob store, and sending out a message that contains an SHA hash of that document. The generation of the PDF document is not 100% deterministic due to a timestamp that is rendered in the top right corner. If we allow the PDF to be stored in a deterministic location, same for each processing attempts, we may end up in a situation where the message that is sent out contains a hash of a document that has been overridden by another processing attempt. This may happen if the incoming message has been duplicated. One copy has been processed correctly while the other copy succeeded to complete the Prepare phase but failed to commit and was subsequently discarded as a duplicate.
```